### PR TITLE
[SPARK-46193][CORE][TESTS] Add `PersistenceEngineBenchmark`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to provide a new benchmark, `PersistenceEngineBenchmark`. 

### Why are the changes needed?

This is beneficial for both the developers and the users by providing a consistent measurement.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manual review.

```
$ build/sbt "core/Test/runMain org.apache.spark.deploy.master.PersistenceEngineBenchmark"
...
[info] OpenJDK 64-Bit Server VM 17.0.9+9-LTS on Mac OS X 14.2
[info] Apple M1 Max
[info] 1000 Workers:                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] ZooKeeperPersistenceEngine                        11179          11198          20          0.0    11179348.5       1.0X
[info] FileSystemPersistenceEngine                         416            422           6          0.0      415745.2      26.9X
[info] BlackHolePersistenceEngine                            0              0           0         22.7          44.1  253597.7X
```

```
$ bin/spark-submit --driver-memory 6g --class org.apache.spark.deploy.master.PersistenceEngineBenchmark --jars `find ~/Library/Caches/Coursier/v1 -name 'curator-test-*.jar'` core/target/scala-2.13/spark-core_2.13-4.0.0-SNAPSHOT-tests.jar
...
OpenJDK 64-Bit Server VM 17.0.9+9-LTS on Mac OS X 14.2
Apple M1 Max
1000 Workers:                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
ZooKeeperPersistenceEngine                        11565          11857         373          0.0    11564757.8       1.0X
FileSystemPersistenceEngine                         426            426           1          0.0      425605.0      27.2X
BlackHolePersistenceEngine                            0              0           0         27.4          36.5  316478.5X
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
I'm going to add the generated results. Currently, it's running.
- https://github.com/dongjoon-hyun/spark/actions/runs/7055533250 (Java 17)
- https://github.com/dongjoon-hyun/spark/actions/runs/7055536305 (Java 21)

File: core/src/test/scala/org/apache/spark/deploy/master/PersistenceEngineBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * {{{
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class> --jars `find ~/.cache/coursier -name 'curator-test-*.jar'` <spark core test jar>
+ *   2. build/sbt "core/Test/runMain <this class>"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "core/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/PersistenceEngineBenchmark-results.txt".
+ * }}}
+ * */
+object PersistenceEngineBenchmark extends BenchmarkBase {
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/deploy/master/PersistenceEngineBenchmark.scala:46]: Thank you for review, @beliefer . Our Github Action `benchmark` job uses the following `spark-submit` command for all benchmarks.

https://github.com/apache/spark/blob/9c6782fff9ec7073281035f1dd0f1273f931d337/.github/workflows/benchmark.yml#L178-L182

**Comment by dongjoon-hyun** [comment]
I added the results. Could you review this once more, @beliefer ?

**Review by beliefer — APPROVED** [review]
LGTM.

**Pull request closed without merging**